clk: axi-clkgen: check axi clock before bus access

Liambeguin opened the pull request:
If an axi read (or write) is performed before the clock is configured,
the driver hangs indefinitely. Make sure the axi clock is ready before
trying to access the bus.

This change forces the s_axi_aclk to be the first clock defined in the
devicetree.

Signed-off-by: Liam Beguin <lvb@xiphos.com>

Liambeguin commented:
Hi, 

I'm opening this PR as a draft because I'd like to find a better way of handling the `parent_names[i]` to not force the order in which the devicetree clocks have to be defined.

I was hoping to use the clock name but I still haven't found a way of accessing the parent name given a "local" clock.

If this is an acceptable patch, I'll update devicetree files accordingly and resend.

here is what I'm using:
```dts
 axi_tx_clkgen: axi-clkgen@83c00000  {
                        compatible = "adi,axi-clkgen-2.00.a";
                        reg = <0x83c00000 0x10000>;
                        #clock-cells = <0>;
                        clocks = <&clk 71>, <&clk0_ad9528 1>;
                        clock-names = "s_axi_aclk", "clkin1";
                        clock-output-names = "axi_tx_clkgen";
};
```

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -566,6 +568,16 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 	if (!axi_clkgen)
 		return -ENOMEM;
 
+	axi_clkgen->axi_clk = devm_clk_get(&pdev->dev, "s_axi_aclk");
```
the approach is definitely correct;
i guess this was missed in this driver for so long, because the probe order was fine and the axi-clkgen never got probed before the ref-clk;

it does need a few tweaks;

going forward, i would change this `s_axi_aclk` to `ref-clk` or `clkin` or something like that;
the ref clock (for the AXI Clock Gen) can be from any source including a clock-chip (like ad9528)

commodo replied:
so, if we use the code i suggested below [https://github.com/analogdevicesinc/linux/pull/1132#discussion_r474534860], we won't need to care about the clock-names at all;
it looks to me that the 2 clocks [in the wiki clk & clk2] are defined as parent clocks in the clock-driver;
so we just need to adapt the logic of the driver to make sure the parent clocks are enabled;

the initial logic you posted is good overall;
but if we're going to make sure that we enable one clock, we should make sure we enable both;
and also, we need to make sure to disable them in the proper sequence as well;

so, do you want to do the final implementation? if not, we can do it; we will definitely ask you to test it :)

Liambeguin replied:
> we won't need to care about the clock-names at all

Oh! okay, I see. I'll implement that and resend.

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -578,12 +590,12 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 
 	init.num_parents = of_clk_get_parent_count(pdev->dev.of_node);
-	if (init.num_parents < 1 || init.num_parents > 2)
+	if (init.num_parents < 1 || init.num_parents > 3)
```
actually, looking into the clock framework a bit,  we may need to find a new way to do it;

it looks like the parent assignment should solve things, but we probably shouldn't do any axi-clk reads in probe;

i'll take a look and come back

commodo replied:
i looked over the kernel;
so, let's drop this parent change; it should be fine as is;

so , the general approach looks fine as is;
i'd propose it like this:
```

struct axi_clkgen {
        void __iomem *base;
        struct clk_hw clk_hw;
        unsigned int pcore_version;
        struct clk *parent_clocks[2];   // maybe let's add this new field
};

static int axi_clkgen_probe(struct platform_device *pdev)
{
....................
        axi_clkgen = devm_kzalloc(&pdev->dev, sizeof(*axi_clkgen), GFP_KERNEL);
        if (!axi_clkgen)
                return -ENOMEM;

        mem = platform_get_resource(pdev, IORESOURCE_MEM, 0);
        axi_clkgen->base = devm_ioremap_resource(&pdev->dev, mem);
        if (IS_ERR(axi_clkgen->base))
                return PTR_ERR(axi_clkgen->base);

        init.num_parents = of_clk_get_parent_count(pdev->dev.of_node);
        if (init.num_parents < 1 || init.num_parents > 2)
                return -EINVAL;

        for (i = 0; i < init.num_parents; i++) {
                axi_clkgen->parent_clocks[i] = of_clk_get(pdev->dev.of_node, i);
                if (IS_ERR(axi_clkgen->parent_clocks[i]) {
                      while (--i >= 0)
                              clk_put(axi_clkgen->parent_clocks[i]);
                       return PTR_ERR(axi_clkgen->parent_clocks[i]);
                }
                parent_names[i] = of_clk_get_parent_name(pdev->dev.of_node, i);
                if (!parent_names[i])
                        return -EINVAL;
        }

        for (i = 0; i < init.num_parents; i++) {
                ret = clk_prepare_enable(axi_clkgen->parent_clocks[i]);
                if (ret < 0) {
                      while (--i >= 0)
                              clk_disable_unprepare(axi_clkgen->parent_clocks[i]);
                       return ret;
                }
        }

      axi_clkgen_read(axi_clkgen, ADI_AXI_REG_VERSION,
                        &axi_clkgen->pcore_version);

        if (ADI_AXI_PCORE_VER_MAJOR(axi_clkgen->pcore_version) > 0x04)
                axi_clkgen_setup_ranges(axi_clkgen);

////////////// from here-on all error-exit paths need to be `goto err_disable_parent_clocks`

....................

err_disable_parent_clocks;
        // this also needs to be done in `axi_clkgen_remove()`
        for (i = 0; i < init.num_parents; i++)
                clk_disable_unprepare(axi_clkgen->parent_clocks[i]);

        return ret;
}
```

commodo commented:
A question: does the issue happen on 2018_R2 as well?

commodo commented:
> > A question: does the issue happen on 2018_R2 as well?
> 
> Yes, I noticed this on a branch based on `2018_R2`

Thanks for confirming this. We had an internal discussion that maybe there was some other patch that broke this, but if you say this happens in 2018_R2, then it's likely an older issue that wasn't noticed before.

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +573,41 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+		if (IS_ERR(axi_clkgen->parent_clocks[i])) {
+			while (--i > 0)
+				clk_put(axi_clkgen->parent_clocks[i]);
+			return -EPROBE_DEFER;
```
i would just return `PTR_ERR(axi_clkgen->parent_clocks[i])` here;
i think `of_clk_get()` would also return `-EPROBE_DEFER` here if that's the case; [i hope that's the case; at least that's what i remember]
and if the case is for another error code, then we should return that;

commodo replied:
i know i suggested the use of `of_clk_get()`, but looking at it now, i'm wondering now if we should use `devm_clk_get(&pdev->dev, parent_names[i])`

that would simplify the exit-on-error code paths; for example no more  need  for `clk_put()`

i am a bit unsure if we need to add `clk_put()` at the `err_disable_parent_clocks` label;
so, when unsure, i'd go for the device-managed version of clk_get() (i.e  devm_clk_get())
that way we don't need to worry about cleaning up clock references;

Liambeguin replied:
I tried to return `PTR_ERR(axi_clkgen->parent_clocks[i])` like you suggested but that fails.
I looked a little into it briefly and it doesn't look like `of_clk_get()` returns `-EPROBE_DEFER`.

Liambeguin replied:
I should've thought about using the `devm_` variant. I'll update

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +573,41 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+		 * and is not an MMCM parent clock
+		 */
+		if (i > 0)
+			parent_names[i - 1] = of_clk_get_parent_name(pdev->dev.of_node, i);
```
i'd leave  `parent_names[i ] = of_clk_get_parent_name(pdev->dev.of_node, i);`
this shifting of indexes looks like it can cause problems with some device-trees; for example you could reverse the order of clocks in the device-tree and then this assumption becomes false [and problematic];

if we try to adhere too much to what's in the Vivado project/topology, we risk complicating things too much here in the driver;
so, if it's not too much of a problem i would vote to keep clock-parent names and clock-parent references in sync;

Liambeguin replied:
I don't like this part either but adding the AXI control clock to the list of parents seems wrong to me. That's why I wanted to use clock names initially

commodo replied:
i wonder if `devm_clk_get(&pdev->dev, parent_names[i])` returns -EPROBE_DEFER?
what is the error code that of_clk_get() returns?

Liambeguin replied:
I'm sure that would work.`devm_clk_get` takes the clock name of the current node and not the parent_name

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +573,41 @@ static int axi_clkgen_probe(struct platform_device *pdev)
-
 	init.num_parents = of_clk_get_parent_count(pdev->dev.of_node);
-	if (init.num_parents < 1 || init.num_parents > 2)
+	if (init.num_parents < 1 || init.num_parents > 3)
```
do we still need to do `if (init.num_parents < 1 || init.num_parents > 3)` ?

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -53,6 +54,7 @@ struct axi_clkgen {
 	void __iomem *base;
 	struct clk_hw clk_hw;
 	unsigned int pcore_version;
+	struct clk *parent_clocks[3];
```
why 3 clocks?

Liambeguin replied:
We have 3 clocks here because we have the 2 MMCM input clocks and the AXI control clock. I added the AXI control clock to `parent_clocks` to have it in the cleanup.

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -601,16 +622,33 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 {
+	int i;
+	int num_parents;
+	struct axi_clkgen *axi_clkgen = platform_get_drvdata(pdev);
```
so, i've developed a style preference here; it's called reverse christmas tree ordering:
```
struct axi_clkgen *axi_clkgen = platform_get_drvdata(pdev);
int num_parents;
int i;
```

i am not fully insisting on this; 
but i know that some upstream maintainers prefer this, so when this would go upstream, we'll probably use this order;

Liambeguin replied:
Okay, I'll update what's left after the switch to the `devm_` variants

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -601,16 +622,33 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 	of_clk_del_provider(pdev->dev.of_node);
 
+	num_parents = of_clk_get_parent_count(pdev->dev.of_node);
+	for (i = 0; i < num_parents; i++)
```
we can also do `for (i = 0; i < ARRAY_SIZE(axi_clkgen->parent_clocks[i])); i++)`
`clk_disable_unprepare()` is tolerant for NULL [if i remember correctly]

Liambeguin commented:
I was confused by the clock name vs. parent clock name.
I've re-implemented this with `clock-names` and I'm happier with how this looks. No more odd offsets... 
I also tried to address your other comments. If this is good, I'll update the documentation and devicetree files and resend.

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +574,33 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 	if (IS_ERR(axi_clkgen->base))
 		return PTR_ERR(axi_clkgen->base);
 
+	for (i = 0, init.num_parents = 0; i < 2; i++) {
```
`i < ARRAY_SIZE(axi_clkgen->parent_clocks[i])` ?

Liambeguin replied:
okay, I'll fix that

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +574,33 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+	for (i = 0, init.num_parents = 0; i < 2; i++) {
+		sprintf(clkin_name, "clkin%d", i + 1);
+		axi_clkgen->parent_clocks[i] = devm_clk_get(&pdev->dev,clkin_name);
+		if (PTR_ERR(axi_clkgen->parent_clocks[i]) == -ENOENT)
```
before, this was enforcing exactly 2 clocks; now it's enforcing a minimum of one;
i'd need to ask some HDL people about this;

commodo replied:
i assumed that `devm_clk_get(&pdev->dev, parent_names[i])` would work;
but i understand now the difference too;

in that case, unfortunately we can't use the `devm_clk_get()` function, as that will break existing device-trees;
so we'd probably be stuck with using `of_clk_get()` ; maybe we don't need to do `clk_put()`
https://github.com/analogdevicesinc/linux/blob/master/drivers/clk/samsung/clk-exynos5433.c#L5533

i keep looking around for how other clock drivers do it, and only one [weird] driver seemed to do it;

Liambeguin replied:
> before, this was enforcing exactly 2 clocks; now it's enforcing a minimum of one

If I'm not mistaken,`if (init.num_parents < 1 || init.num_parents > 2)` took one or two clocks.

> in that case, unfortunately we can't use the devm_clk_get() function, as that will break existing device-trees

My intention was to update the devicetrees (and documentation) as soon as we've settled on the implementation

commodo replied:
> If I'm not mistaken,`if (init.num_parents < 1 || init.num_parents > 2)` took one or two clocks.

oh right; my bad;


> My intention was to update the devicetrees (and documentation) as soon as we've settled on the implementation

my fear here is that we update ours, but other people using our kernel [or the upstream kernel] don't update their device-trees and then https://ez.analog.com/ rings-up like crazy;
so, whatever we can do to fix bugs, but keep the old behavior working, is more important;
( fwiw: i don't like doing https://ez.analog.com/ work :p  )

Liambeguin replied:
it can be hard to figure out what's happening with these drivers but I really appreciate the support on ez.analog.com!

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -571,22 +574,33 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+		if (IS_ERR(axi_clkgen->parent_clocks[i]))
+			return PTR_ERR(axi_clkgen->parent_clocks[i]);
+
+		ret = clk_prepare_enable(axi_clkgen->parent_clocks[i]);
```
i'd split this `clk_prepare_enable()` in another loop; (maybe after `if (init.num_parents < 1) {`)
the logic here doesn't seem to disable clocks in all cases;
for example; in the `(init.num_parents < 1)` check below, clocks aren't disabled;
and in the `if (IS_ERR(axi_clkgen->parent_clocks[i]))` check above, the first clock may be left enabled, if the 2nd fails;

Liambeguin replied:
sorry about that, I'll rework that and split the loop into 2

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -574,6 +575,16 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 		return PTR_ERR(axi_clkgen->base);
 
+	axi_clkgen->axi_clk = devm_clk_get(&pdev->dev, "s_axi_aclk");
+	if (IS_ERR(axi_clkgen->axi_clk)) {
```
i'm a bit puzzled about this `s_axi_aclk` ; can't this be one of the 2 `clkin1` or `clkin2`?
do we really need to enforce this name?
in some other setups for this clk-gen, maybe it's not  any AXI clock, but maybe an input from a clock chip (AD9523 or AD9528)?

Liambeguin replied:
The `s_axi_aclk` is the clock needed to perform read/write operations on the AXI bus. It is not used with the MMCM and should not be set as a parent, so it can't be `clkin1` or `clkin2`.
This is the clock that was the source of my initial issue. In some cases, the clock is not set when the node is probed and the driver hangs on the first call to `axi_clkgen_read`.
This clock has to be setup in Vivado in all cases otherwise the AXI interface cannot be used.

commodo replied:
apologies here for being difficult [or just plain ignorant/stupid];
i rarely open up Vivado [since Linux takes up most of my days];
talked to @Csomi and he confirmed that the AXI clock should be enabled; 
we were a bit puzzled at first about needing this to be enabled, because most of reference systems already have this AXI enabled before Linux boots; we admit that there are some scenarios [which we probably don't cover], where the PS is enabled, but the PL is enabled at a later stage;

so this change is good;

coming back to my previous comment here https://github.com/analogdevicesinc/linux/pull/1132/commits/bcea2565719f665b2f170bb339bd88fd619f2d01#r478266551
i think we will need to keep the `clkin1` and `clkin2` names; so that change is also fine;
we will have to update all our DTs, and deal with EZ (which... I guess we'll have to; the best way is not always the easiest I guess)

i know i seem to be changing my mind, but that's because i seem to not have the full understanding of things [as you do]

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -574,6 +575,16 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+		return PTR_ERR(axi_clkgen->axi_clk);
+	}
+
+	ret = clk_prepare_enable(axi_clkgen->axi_clk);
```
this clock should be disabled in probe as well in the same place the parent clocks are disabled [on the error path];

commodo commented:
hopefully i have a better understanding of all things;
so this patchset is fine overall;
let's refine these and then update the device-trees;

Liambeguin commented:
I reworked the probe function based on your comments and updated the devicetree files and the Documentation.
I had to assume what the `s_axi_aclk` clock was:
 - for `arm`, I set it to `fclk0`
 - for `arm64`, I set it to `pl0_ref`
 - for `microblaze`, I set it to `clk_bus_0`

These should be checked in the different vivado projects.

Another note, disabling the `s_axi_aclk` when the probe fails might make other drivers fail down the line if they don't check for it too (which they should).

commodo commented:
there's some issues with some device-trees:
https://travis-ci.org/github/analogdevicesinc/linux/jobs/721870078#L556

you can also run this check locally, if you run ` DTS_FILES="arch/arm/boot/dts/zynq-*.dts arch/arm/boot/dts/socfpga_*.dts arch/arm64/boot/dts/xilinx/zynqmp-*.dts arch/microblaze/boot/dts/*.dts arch/nios2/boot/dts/*.dts" ci/travis/run-build.sh dtb_build_test`

would it be too much to ask to put the device-tree changes in separate patches from the driver changes?
i.e. 1) `clkin1` dt changes    2) driver change for parent-clocks-names   3) axi_clk dt changes  4)   axi_clk  driver changes

we will also want to upstream these driver changes , so if you do it like this, it will be easier to cherry-pick them;

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -641,6 +656,8 @@ static int axi_clkgen_remove(struct platform_device *pdev)
 	for (i = 0; i < ARRAY_SIZE(axi_clkgen->parent_clocks); i++)
 		clk_disable_unprepare(axi_clkgen->parent_clocks[i]);
 
+	clk_disable_unprepare(axi_clkgen->axi_clk);
```
there's a general rule about `remove` & `exit-on-error`; the order should be the same;
here it's not particularly important, but when this goes upstream, we would need to do it;

so, one option is:
```
// we keep this as is
err_disable_axi_clk:
	clk_disable_unprepare(axi_clkgen->axi_clk);

err_disable_parent_clocks:
	for (i = 0; i < init.num_parents; i++)
		clk_disable_unprepare(axi_clkgen->parent_clocks[i]);
```

and change this:
```
	clk_disable_unprepare(axi_clkgen->axi_clk);

	for (i = 0; i < ARRAY_SIZE(axi_clkgen->parent_clocks); i++)
		clk_disable_unprepare(axi_clkgen->parent_clocks[i]);

```

or do it the other way around;
sometimes this order is very important as you should disable things in the mirrored order that was initialized;
so [for example] if a voltage regulator is enabled first [to provide power to the device], it should be disabled last, to make sure we don't cut power to the device before doing any proper shutdown;

Liambeguin commented:
Sorry about the bad devicetrees.. I updated the files and split the commits like you suggested.
I left the Documentation changes with the axi clk commit.

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -574,12 +575,6 @@ static int axi_clkgen_probe(struct platform_device *pdev)
 	if (IS_ERR(axi_clkgen->base))
 		return PTR_ERR(axi_clkgen->base);
 
-	axi_clkgen_read(axi_clkgen, ADI_AXI_REG_VERSION,
```
this isn't an issue yet, but would it be a problem to leave the `axi_clkgen_read()` here? and add the axi_clk setup before this part?

it would also mean that `clk_disable_unprepare(axi_clkgen->axi_clk);` would go after disabling the parent-clocks [during remove & the error path], but that would actually make some sense, since [in some potential future versions of this driver] someone would want to squeeze some register reads/writes before

Liambeguin replied:
I'm not sure I understand this comment, is this on the latest revision I sent?
On the diff I see, the axi_clkgen_read is right after the `s_axi_aclk` setup.

commodo replied:
the suggestion is to add here, this block:
```
	axi_clkgen->axi_clk = devm_clk_get(&pdev->dev, "s_axi_aclk");
	if (IS_ERR(axi_clkgen->axi_clk)) {
		dev_err(&pdev->dev, "failed to get s_axi_aclk\n");
		return PTR_ERR(axi_clkgen->axi_clk);
	}
```

naturally `clk_disable_unprepare(axi_clkgen->axi_clk);` would need to be done after the parent-clocks disable-unprepare;
and the error-paths of the parent-clocks init would need to be updated;

this change can be done now, or later; either is fine from my side;
it's the only remaining comment i have for this set;

Liambeguin replied:
Hi @commodo 

I'm sorry if I'm still confused but in the latest revision I sent, the `devm_clk_get(&pdev->dev, "s_axi_aclk")` and the `axi_clkgen_read` were moved after the clock setup code (https://github.com/Liambeguin/linux-1/blob/axi-clkgen/drivers/clk/clk-axi-clkgen.c#L615).

So right now, the first `axi_clkgen_read` happens after the clock setup. Is that what you meant?

commodo replied:
i'm saying that we could check for the axi_clk first and fail early [and not enable any parent clocks] if the axi_clk is not ready/available;

Liambeguin replied:
Oh! sorry for not getting that earlier... I'll update and resend

commodo commented on drivers/clk/clk-axi-clkgen.c:
```diff
@@ -577,14 +580,27 @@ static int axi_clkgen_probe(struct platform_device *pdev)
+		parent_names[i] = __clk_get_name(axi_clkgen->parent_clocks[i]);
+		init.num_parents++;
+	}
+
+	if (init.num_parents < 1) {
+		dev_err(&pdev->dev, "Missing input clock, see 'clkin1' and 'clkin2'\n");
 		return -EINVAL;
+	}
 
 	for (i = 0; i < init.num_parents; i++) {
-		parent_names[i] = of_clk_get_parent_name(pdev->dev.of_node, i);
-		if (!parent_names[i])
-			return -EINVAL;
+		ret = clk_prepare_enable(axi_clkgen->parent_clocks[i]);
```
the clock framework requires that all clk_prepare_enable /  clk_disable_unprepare  calls be balanced;
so for each enable, there should be a symmetrical disable on each exit/error path;
that's because it keeps some internal refcounts; so if you do 1 enable and 2 disables [on a clock], you'd get a kernel warning/stack-trace message; and if you do 1 enable call, and 0 disable calls, the clock may never be disabled [gracefully] on system powerdown;

anyway, in this case, for this loop, this shouldn't do a `goto err_disable_parent_clocks`, because [for 2 clocks], this would call `clk_disable_unprepare()` for both clocks; in case the first one is enabled [successfully] and the 2nd isn't, you'd get a refcount warning from the clock framework, that you are trying to disable a clock that wasn't enabled;

Liambeguin replied:
Understood, thanks for the explanation. I'll use the loop you suggested initially.

commodo commented:
this is something i should have researched better in the beginning;

> Sorry about the bad devicetrees.. I updated the files and split the commits like you suggested.
> I left the Documentation changes with the axi clk commit.

no worries about the DTs; we've broken plenty of them as well, and that's why we added the CI in place to catch it for us;

the current changeset looks pretty good now;
only 2 comments;

commodo commented:
many thanks for sitting through this one :)

Liambeguin commented:
No worries, thanks for your time and support :)
